### PR TITLE
feat(source): create source with connection/drop connection

### PR DESCRIPTION
**Pull request by WillyKidd**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

- Moved `Connection` from root catalog to schema catalog
- Refined the `Create Connection` and `Show Connections` statements according to the changes
- Implemented the `Drop Connection` statement
- Added `MockConnection` connection type for testing
- Added some e2e tests for the above features

This is a followup of the last pr: https://github.com/risingwavelabs/risingwave/pull/8907/

Related: https://github.com/risingwavelabs/risingwave/issues/8771, https://github.com/risingwavelabs/risingwave/pull/9119

## Checklist For Contributors

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [x] I have demonstrated that backward compatibility is not broken by breaking changes and created issues to track deprecated features to be removed in the future. (Please refer to the issue)
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Checklist For Reviewers

- [x] I have requested macro/micro-benchmarks as this PR can affect performance substantially, and the results are shown.

## Documentation

- [x] My PR **DOES** contain user-facing changes.

### Types of user-facing changes

- SQL commands, functions, and operators

### Release note

- Added support for `CREATE SOURCE` with a privatelink connection. Syntax:
  ```sql
  CREATE SOURCE tcp_metrics (
      device_id VARCHAR,
      metric_name VARCHAR,
      report_time TIMESTAMP,
      metric_value DOUBLE PRECISION
  ) WITH (
      connector = 'kafka',
      topic = 'tcp_metrics',
      scan.startup.mode = 'earliest',
      properties.bootstrap.server = 'b-1.test.ra68za.c10.kafka.us-east-1.amazonaws.com:9092,b-2.test.ra68za.c10.kafka.us-east-1.amazonaws.com:9092',
      connection.name = 'connection_name',
      privatelink.targets = '[{"port": 8001}, {"port": 8002}]',
  ) ROW FORMAT JSON;
  ```
- Added support for `DROP CONNECTION`. Syntax:
  ```sql
  DROP CONNECTION {{ name }};
  ```

### CODE REVIEW

File: e2e_test/ddl/connection.slt [hunk 1]
```diff
@@ -0,0 +1,19 @@
+# Create a connection.
+statement ok
+CREATE CONNECTION conn0 WITH (type = 'mock');
+
+# Create another user with duplicate name.
+statement error
+CREATE CONNECTION conn0 WITH (type = 'mock');
```
Comment (StrikeW) [inline e2e_test/ddl/connection.slt:7]: Add some cases for `IF NOT EXISTS`

File: src/meta/src/manager/catalog/database.rs [hunk 2]
```diff
@@ -185,6 +199,18 @@ impl DatabaseManager {
         }
     }
 
+    pub fn check_connection_name_duplicated(&self, relation_key: &RelationKey) -> MetaResult<()> {
```
Comment (yezizp2012) [inline src/meta/src/manager/catalog/database.rs:202]: Yes, we should only check if it's duplicate in the connection collection. But in frontend it's checked among all relations.

File: src/meta/src/rpc/service/ddl_service.rs [hunk 3]
```diff
@@ -673,10 +697,11 @@ where
 
     async fn resolve_private_link_info(
         &self,
+        connection_id: ConnectionId,
         properties: &mut HashMap<String, String>,
     ) -> MetaResult<()> {
         let mut broker_rewrite_map = HashMap::new();
-        const UPSTREAM_SOURCE_PRIVATE_LINK_KEY: &str = "private.links";
+        const UPSTREAM_SOURCE_PRIVATE_LINK_KEY: &str = "privatelink.targets";
```
Comment (yezizp2012) [inline src/meta/src/rpc/service/ddl_service.rs:704]: Why renaming here? Prefer the former one.

**Review by yezizp2012 — COMMENTED** [review]
Rest LGTM

File: src/meta/src/rpc/service/ddl_service.rs (hunk 3 above)
Comment (WillyKidd) [reply src/meta/src/rpc/service/ddl_service.rs:704]: @StrikeW what do you think?

File: src/frontend/src/session.rs [hunk 4]
```diff
@@ -506,10 +507,31 @@ impl SessionImpl {
...
+            let schema_name = match schema_name {
+                Some(schema_name) => schema_name,
+                None => catalog_reader
+                    .first_valid_schema(db_name, &search_path, user_name)?
+                    .name(),
+            };
+            (schema_name, connection_name)
+        };
+        catalog_reader
+            .check_relation_name_duplicated(db_name, &schema_name, &connection_name)
```
[the remark this answers is not in the document]
Comment (WillyKidd) [reply src/frontend/src/session.rs:534]: oooops! my mistake 🤣

File: src/meta/src/manager/catalog/mod.rs [hunk 5]
```diff
@@ -358,44 +350,64 @@ where
...
     }
 
-    pub async fn drop_connection(&self, conn_name: &str) -> MetaResult<NotificationVersion> {
-        let core = &mut self.core.lock().await.connection;
-
-        let conn_id = core
-            .connection_by_name
-            .remove(conn_name)
-            .ok_or_else(|| anyhow!("connection {} not found", conn_name))?;
+    pub async fn drop_connection(&self, conn_id: ConnectionId) -> MetaResult<NotificationVersion> {
```
Comment (yezizp2012) [inline src/meta/src/manager/catalog/mod.rs:380]: Oh, we should also drop all connections under a database when receiving `drop database` command. See `drop_database` for more details.

File: proto/catalog.proto [hunk 6]
```diff
@@ -90,8 +91,9 @@ message Sink {
   string definition = 13;
 }
 
-// TODO: add database_id and schema_id
 message Connection {
+  message MockConnection {}
```
[the remark this answers is not in the document]
Comment (StrikeW) [reply proto/catalog.proto:95]: Yes. I think we can reuse the `provider` field to achieve the same goal.

File: src/meta/src/rpc/service/ddl_service.rs (hunk 3 above)
Comment (StrikeW) [reply src/meta/src/rpc/service/ddl_service.rs:704]: Let's keep the newer one. The previous one doesn't expose the concept of `Target`, which is the destination where RisingWave needs to route the network traffic.

File: e2e_test/ddl/connection.slt [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+CREATE CONNECTION conn0 WITH (type = 'privatelink', provider = 'mock');
+
+# Create another user with duplicate name.
+statement error
+CREATE CONNECTION conn0 WITH (type = 'privatelink', provider = 'mock');
+
+# Create if not exists.
+statement ok
+CREATE CONNECTION IF NOT EXISTS conn0 WITH (type = 'privatelink', provider = 'mock');
+
```
Comment (StrikeW) [inline e2e_test/ddl/connection.slt:12]: Should add a case for `conn0` is already existed.

File: src/meta/src/rpc/service/ddl_service.rs [hunk 8]
```diff
@@ -596,48 +600,77 @@ where
...
+                    PbPrivateLinkProvider::Mock => {
+                        let id = self.gen_unique_id::<{ IdCategory::Connection }>().await?;
+                        let private_link_svc = PbPrivateLinkService {
+                            provider: link.provider,
+                            service_name: String::new(),
+                            endpoint_id: String::new(),
+                            endpoint_dns_name: String::new(),
+                            dns_entries: HashMap::new(),
+                        };
+                        let connection = Connection {
```
Comment (StrikeW) [inline src/meta/src/rpc/service/ddl_service.rs:621]: There are duplicate code in these two branches, why not remove the redundancy?

**Review by StrikeW — APPROVED** [review]
LGTM

File: src/meta/src/manager/catalog/mod.rs [hunk 9]
```diff
@@ -1266,13 +1284,22 @@ where
         } else {
             database_core.mark_creating(&key);
             user_core.increase_ref(source.owner);
+            if let Some(connection_id) = source.connection_id {
```
Comment (yezizp2012) [inline src/meta/src/manager/catalog/mod.rs:1287]: Need to ensure existence of the connection id, as the connection could be dropped from other frontend at the same time.

**Review by yezizp2012 — APPROVED** [review]
Rest LGTM.

File: src/meta/src/manager/catalog/mod.rs (hunk 9 above)
Comment (WillyKidd) [reply src/meta/src/manager/catalog/mod.rs:1287]: wow! I didnt think of that! so considerate!

**Pull request merged**